PWGHF/D2H/Tasks/taskDplus.cxx

```diff
@@ -307,53 +500,220 @@ struct HfTaskDplus {
         if ((yCandRecoMax >= 0. && std::abs(hfHelper.yDplus(candidate)) > yCandRecoMax)) {
           continue;
         }
+        ptBhad = candidate.ptBhadMotherPart();
+        flagBHad = getBHadMotherFlag(candidate.pdgBhadMotherPart());
+
+        if (storeCentrality || storeOccupancy) {
+          auto collision = candidate.template collision_as<McRecoCollisionsCent>();
+          if (storeCentrality && centEstimator != CentralityEstimator::None) {
+            cent = getCentrality(collision);
+          }
+          if (storeOccupancy && occEstimator != OccupancyEstimator::None) {
+            occ = getOccupancy(collision);
+          }
+        }
+
         fillHisto(candidate);
         fillHistoMCRec<true>(candidate);
-        fillSparseML<true, true>(candidate);
+        fillSparseML<true, true>(candidate, ptBhad, flagBHad, cent, occ);
       }
       // Bkg
+      ptBhad = -1;
+      flagBHad = -1;
       for (const auto& candidate : recoBkgCandidatesWithMl) {
         if ((yCandRecoMax >= 0. && std::abs(hfHelper.yDplus(candidate)) > yCandRecoMax)) {
           continue;
         }
+        auto collision = candidate.template collision_as<McRecoCollisionsCent>();
+        if (storeCentrality && centEstimator != CentralityEstimator::None) {
+          cent = getCentrality(collision);
+        }
+        if (storeOccupancy && occEstimator != OccupancyEstimator::None) {
+          occ = getOccupancy(collision);
+        }
         fillHistoMCRec<false>(candidate);
-        fillSparseML<true, false>(candidate);
+        fillSparseML<true, false>(candidate, ptBhad, flagBHad, cent, occ);
       }
     }
+  }
+
+  // Run analysis for the generated Dplus candidates
+  /// \param mcGenCollisions are the generated MC collisions
+  /// \param mcRecoCollisions are the reconstructed MC collisions
+  /// \param mcGenParticles are the generated MC particle candidates
+  template <bool fillMl, typename Cand>
+  void runMCGenAnalysis(aod::McCollisions const& mcGenCollisions,
+                        McRecoCollisionsCent const& mcRecoCollisions,
+                        const Cand& mcGenParticles)
```

Comment:
```suggestion
                        Cand const& mcGenParticles)
```

Reply:
Aren't the two expressions equivalent? What is the advantage of the second option?

Reply:
Readability and consistency.